Introduce Internal Test Artifact Plugin

breskeby opened the pull request:
This reduces the ceremony declaring test artifacts for a project.

The change also solves an issue with usage of deprecated `testRuntime` that
testArtifacts extendsFrom which seems not required at all and will break with Gradle 7.0

breskeby committed:
Introduce Internal Test Artifact Plugin

This reduces the ceremony declaring test artifacts for a project.
This also solves an issue with usage of deprecated testRuntime that
testArtifacts extendsFrom which seems not required at all and would have
broke with Gradle 7.0 anyhow

elasticmachine commented:
Pinging @elastic/es-delivery (Team:Delivery)

pugnascotia commented on buildSrc/src/main/java/org/elasticsearch/gradle/internal/InternalTestArtifactPlugin.java:
```diff
@@ -0,0 +1,39 @@
+import org.gradle.api.tasks.bundling.Jar;
+
+/**
+ * Ideally, yhis plugin is intended to be temporary and in the long run we want to move
```
```suggestion
 * Ideally, this plugin is intended to be temporary and in the long run we want to move
```

breskeby replied:
fixed manually

pugnascotia reviewed (approved):
LGTM, which is to say, I don't know what these tasks are doing with the test jar, but the replacement with a dedicated plugin looks good.

nik9000 commented on qa/full-cluster-restart/build.gradle:
```diff
@@ -58,18 +59,4 @@ for (Version bwcVersion : BuildParams.bwcVersions.indexCompatible) {
   tasks.register(bwcTaskName(bwcVersion)) {
     dependsOn tasks.named("${baseName}#upgradedClusterTest")
   }
-}
```
Because these are purely test projects I wonder if we could just make the changes we want here in the end? I get being stuck with test artifacts for things like `server` because no one has time to untangle it. But for this one and the next one maybe its pretty simple?

I guess it doesn't make sense to block the change on this though. But maybe it'd be worth doing for the easy ones

breskeby replied:
We can look into this in the future, but I would not want to mix this into this PR

nik9000 reviewed (approved):
I like centralizing the sadness. I think we could fix this for the qa projects pretty quickly, but we're stuck with it for the x-pack submodules I think. At least for a long while.

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/internal/InternalTestArtifactPlugin.java:
```diff
@@ -0,0 +1,39 @@
+        });
+
+        project.getArtifacts().add("testArtifacts", testJar);
+        project.getPlugins()
```
Why do we need to publish main jar as well? I feel like the `runtimeElements` bit of `testArtifacts` should include that already. Perhaps there's a bit of wiring missing here? I would expect that `testArtifacts` configuration would include with it all the runtime dependencies of the `test` sourceset, to include the main sourceset output.

breskeby replied:
The problem is that the jar is not wired into the test compile configuration but rather added via sourceSets.main.getOutput directly to the compile class path.

mark-vieira reviewed (approved):
One comment, otherwise LGTM. This is a very long overdue refactoring.

breskeby committed:
Extend testRuntimeClasspath for simplication

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/internal/InternalTestArtifactPlugin.java:
```diff
@@ -0,0 +1,34 @@
+    @Override
+    public void apply(Project project) {
+        Configuration testArtifactsConfiguration = project.getConfigurations().create("testArtifacts");
+        testArtifactsConfiguration.extendsFrom(project.getConfigurations().getByName("testRuntimeClasspath"));
```
Would `testRuntimeElements` be more appropriate here? Is there really a difference?

breskeby replied:
I think I'm ending up redoing this by providing a variant aware solution. because that's what we really want here and what is the only _correct_ solution here imo. 
We want the testApiElements at compile time and testRuntimeElements at runtime. 
Also just using testRuntimeElements would bring us some other issue down the road where runtimeOnly elements would end up on test compile classpaths.

breskeby committed:
Make test artifact plugin variant aware

This allows us a more adequate compile and runtime classpath for the
consuming projects.

TODO: Introduce some syntax sugar for consuming projects

breskeby committed:
Minor cleanup on test artifact usage declarations

breskeby committed:
Fix some more testArtifacts usages

breskeby committed:
Fix typo in capability declaration

breskeby committed:
Change test artifact naming pattern to be PolicyUtil compliant

breskeby committed:
Introduce convention method to declare test artifact dependencies

breskeby committed:
Some cleanup on dependencies

breskeby committed:
Add some explanatory code comment to InternalTestArtifactPlugin

breskeby commented on buildSrc/src/main/java/org/elasticsearch/gradle/internal/InternalTestArtifactPlugin.java:
```diff
@@ -0,0 +1,52 @@
+        Configuration testApiElements = project.getConfigurations().getByName("testApiElements");
+        testApiElements.extendsFrom(project.getConfigurations().getByName("testCompileClasspath"));
+        DependencyHandler dependencies = project.getDependencies();
+        project.getPlugins().withType(JavaPlugin.class, javaPlugin -> {
```
the very own project dependency is not added to any test configuration as internally only the direct output of the main sourceSet is added to test dependencies setup. Therefore we to manually add it to the exposed configurations consuming projects will rely on when declaring a test artifact dependency.

breskeby committed:
Use ProviderFactory instead of Project.provider

breskeby commented:
I did some rework since last review to make this variant aware and not let runtime class path leak into compile classpaths

breskeby commented:
@elasticmachine update branch

breskeby commented:
@elasticmachine test this

pugnascotia commented on x-pack/plugin/security/qa/security-basic/build.gradle:
```diff
@@ -4,9 +4,8 @@ apply plugin: 'elasticsearch.java-rest-test'
 import org.elasticsearch.gradle.info.BuildParams
 
 dependencies {
-  javaRestTestImplementation project(path: xpackModule('core'))
```
Other modules have kept this dependency - was it not required?

breskeby replied:
There was an inconsistency in those dependencies by testArtifacts have provided that dependency before. We now ship this dependency on the main artifact as part of the testArtifacts (analog to what the gradle build in testFixture does). 

So when having a dependency on the `testArtifacts(project(path: xpackModule('core')))` we can consistently remove the explicit dependency on the main artifact. I'll go ahead and clean that up

pugnascotia commented on x-pack/plugin/vectors/build.gradle:
```diff
@@ -12,6 +12,5 @@ archivesBaseName = 'x-pack-vectors'
   compileOnly project(path: xpackModule('core'))
-  testImplementation project(path: xpackModule('core'), configuration: 'testArtifacts')
-}
+testImplementation(testArtifact(project(xpackModule('core'))))}
```
Formatting is off here.

breskeby replied:
fixed

pugnascotia commented on x-pack/plugin/voting-only-node/build.gradle:
```diff
@@ -9,5 +9,4 @@ esplugin {
   compileOnly project(path: xpackModule('core'))
-  testImplementation project(path: xpackModule('core'), configuration: 'testArtifacts')
-}
+testImplementation(testArtifact(project(xpackModule('core'))))}
```
Formatting is off here.

breskeby replied:
fixed

pugnascotia commented on x-pack/plugin/wildcard/build.gradle:
```diff
@@ -10,6 +10,5 @@ archivesBaseName = 'x-pack-wildcard'
   compileOnly project(path: xpackModule('core'))
-  testImplementation project(path: xpackModule('core'), configuration: 'testArtifacts')
-}
+testImplementation(testArtifact(project(xpackModule('core'))))}
```
Formatting is off here.

breskeby replied:
fixed

pugnascotia commented on x-pack/qa/evil-tests/build.gradle:
```diff
@@ -1,8 +1,7 @@
 dependencies {
-  testImplementation project(path: xpackModule('security'), configuration: 'testArtifacts')
-}
+  testImplementation(testArtifact(project(xpackModule('security'))))}
```
Formatting is off here.

breskeby replied:
fixed

breskeby replied:
fixed

pugnascotia commented on x-pack/qa/security-setup-password-tests/build.gradle:
```diff
@@ -5,8 +5,7 @@ apply plugin: 'elasticsearch.rest-test'
   testImplementation project(path: xpackModule('security'))
-  testImplementation project(path: xpackModule('core'), configuration: 'testArtifacts')
-}
+testImplementation(testArtifact(project(xpackModule('core'))))}
```
Formatting is off here.

breskeby replied:
fixed

breskeby replied:
fixed

breskeby committed:
Address review feedback

- fix formatting
- fix inconsistency in dependency declarations

breskeby commented:
@pugnascotia addressed your feedback
